File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    
    static int answer = 0;
    static HashMap<String, Integer> list = new HashMap<>();
    static HashMap<String, Integer> sale = new HashMap<>();
    
    public int solution(String[] want, int[] number, String[] discount) {
        
        for(int i=0; i<want.length; i++){
            list.put(want[i], number[i]);
        }
    
        for(int i=0; i<10; i++){
            sale.put(discount[i], sale.getOrDefault(discount[i], 0) + 1);
        }
        
        checkWant(want);
        
        for(int i=10; i<discount.length; i++){
            sale.put(discount[i-10], sale.get(discount[i-10]) -1);
            sale.put(discount[i], sale.getOrDefault(discount[i], 0) + 1);
            checkWant(want);
        }
        
        
        return answer;
    }
    
    public void checkWant(String[] want){
        
        int cnt = 0;
        
        for(int i=0; i<want.length; i++){
            if(list.get(want[i]) <= sale.getOrDefault(want[i], 0)) cnt++;
        }
        
        if(cnt == list.size()) answer++;
        
    }

}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
const list = new Map();
const sale = new Map();
var answer = 0;

function solution(want, number, discount) {
   
    for(let i=0; i<want.length; i++){
        list.set(want[i], number[i]);
    }
    
    for(let i=0; i<10; i++){
        sale.set(discount[i], (sale.get(discount[i]) ?? 0) + 1);
    }
    
    checkWant(want);
    
    for(let i=10; i<discount.length; i++){
        sale.set(discount[i-10], sale.get(discount[i-10]) -1);
        sale.set(discount[i], (sale.get(discount[i]) ?? 0) + 1);
        checkWant(want);
    }
    
    return answer;
}

function checkWant(want){
    
    let cnt = 0;
    
    for(let i=0; i<want.length; i++){
        if(list.get(want[i]) <= (sale.get(want[i]) ?? 0)) cnt++;
    }

    if(cnt == want.length) answer++;
    
}